prov/rxm/src/rxm_cq.c

```diff
@@ -362,49 +326,92 @@ ssize_t rxm_cq_handle_data(struct rxm_rx_buf *rx_buf)
+							 rx_buf->recv_entry->rxm_iov.iov,
+							 rx_buf->recv_entry->rxm_iov.count,
+							 FI_READ, rx_buf->mr);
+				if (ret)
+					return ret;
+
+				for (i = 0; i < rx_buf->recv_entry->rxm_iov.count; i++)
+					rx_buf->recv_entry->rxm_iov.desc[i] =
+						fi_mr_desc(rx_buf->mr[i]);
+			} else {
+				for (i = 0; i < rx_buf->recv_entry->rxm_iov.count; i++)
+					rx_buf->recv_entry->rxm_iov.desc[i] =
+						fi_mr_desc(rx_buf->recv_entry->rxm_iov.desc[i]);
+			}
 
-			for (i = 0; i < rx_buf->recv_entry->rxm_iov.count; i++)
-				rx_buf->recv_entry->rxm_iov.desc[i] =
-					fi_mr_desc(rx_buf->mr[i]);
-		} else {
-			for (i = 0; i < rx_buf->recv_entry->rxm_iov.count; i++)
-				rx_buf->recv_entry->rxm_iov.desc[i] =
-					fi_mr_desc(rx_buf->recv_entry->rxm_iov.desc[i]);
+			/* Ignore the case when the posted recv buffer is not large enough,
+			 * FI_TRUNC error will be generated to user at the end */
+			rxm_match_rma_iov(rx_buf->recv_entry, rx_buf->rma_iov, rx_buf->match_iov);
+			RXM_LOG_STATE_RX(FI_LOG_CQ, rx_buf, RXM_LMT_READ);
+			rx_buf->hdr.state = RXM_LMT_READ;
+			return rxm_lmt_rma_read(rx_buf);
+		case ofi_ctrl_seg_data:
+			done_len = ofi_copy_to_iov(rx_buf->recv_entry->rxm_iov.iov,
+						   rx_buf->recv_entry->rxm_iov.count,
+						   0, rx_buf->pkt.data,
+						   rx_buf->pkt.ctrl_hdr.seg_size);
+			rx_buf->recv_entry->total_recv_len += done_len;
+
+			if ((rx_buf->pkt.ctrl_hdr.seg_no) &&
+			    (done_len == rx_buf->pkt.ctrl_hdr.seg_size)) {
+				rx_buf->recv_entry->last_recv_seg_num =
+							rx_buf->pkt.ctrl_hdr.seg_no;
+				if (rx_buf->recv_entry->msg_id == UINT64_MAX) {
+					rx_buf->recv_entry->msg_id =
+							rx_buf->pkt.ctrl_hdr.msg_id;
+					if (rbtInsert(rx_buf->ep->seg_recv_entry_map,
+						      &rx_buf->recv_entry->msg_id,
+						      rx_buf->recv_entry))
+						goto fn;
+				}
+
+				/* The RX buffer can be reposted for further re-use */
+				rx_buf->recv_entry = NULL;
+				dlist_insert_tail(&rx_buf->repost_entry,
+						  &rx_buf->ep->repost_ready_list);
+				return FI_SUCCESS;
+			}
+			rbtErase(rx_buf->ep->seg_recv_entry_map,
+				 rx_buf->recv_entry->rbt_iter);
+fn:
+			/* Mark rxm_recv_entry::msg_id as unknown for futher re-use */
+			rx_buf->recv_entry->msg_id = UINT64_MAX;
+			return rxm_finish_recv(rx_buf,
+					       rx_buf->recv_entry->total_recv_len);
+		default:
+			assert(0);
+			return -FI_EOTHER;
 		}
-
-		/* Ignore the case when the posted recv buffer is not large enough,
-		 * FI_TRUNC error will be generated to user at the end */
-		rxm_match_rma_iov(rx_buf->recv_entry, rx_buf->rma_iov, rx_buf->match_iov);
-		RXM_LOG_STATE_RX(FI_LOG_CQ, rx_buf, RXM_LMT_READ);
-		rx_buf->hdr.state = RXM_LMT_READ;
-		return rxm_lmt_rma_read(rx_buf);
-	} else {
-		uint64_t done_len = ofi_copy_to_iov(rx_buf->recv_entry->rxm_iov.iov,
-						    rx_buf->recv_entry->rxm_iov.count,
-						    0, rx_buf->pkt.data,
-						    rx_buf->pkt.hdr.size);
-		return rxm_finish_recv(rx_buf, done_len);
 	}
 }
```

Comment:
This function is getting too long and should be broken up.  I think the patch can be restructured to minimize the diffs in this function.